edb/server/protocol/auth_ext/webauthn.py

```diff
@@ -0,0 +1,232 @@
+            credential=credentials,
+            expected_challenge=registration_challenge.challenge,
+            expected_rp_id=self.provider.relying_party_id,
+            expected_origin=self.provider.relying_party_origin,
+        )
+
+        try:
+            result = await execute.parse_execute_json(
+                self.db,
+                """
+with
+    email := <str>$email,
+    user_handle := <str>$user_handle,
+    credential_id := <str>$credential_id,
+    public_key := <str>$public_key,
+    identity := (insert ext::auth::LocalIdentity {
+        issuer := "local",
+        subject := "",
+    }),
+    factor := (insert ext::auth::WebAuthnFactor {
+        email := <str>$email,
+        user_handle := enc::base64_decode(<str>$user_handle),
+        credential_id := enc::base64_decode(<str>$credential_id),
+        public_key := enc::base64_decode(<str>$public_key),
+        identity := identity,
+    }),
+select identity { * };""",
+                variables={
+                    "email": email,
+                    "user_handle": base64.b64encode(user_handle).decode(),
+                    "credential_id": base64.b64encode(
+                        registration_verification.credential_id
+                    ).decode(),
+                    "public_key": base64.b64encode(
+                        registration_verification.credential_public_key
+                    ).decode(),
+                },
+                cached_globally=True,
+            )
+        except Exception as e:
+            exc = await execute.interpret_error(e, self.db)
+            if isinstance(exc, ConstraintViolationError):
+                raise errors.UserAlreadyRegistered()
+            else:
+                raise exc
+
+        result_json = json.loads(result.decode())
+        assert len(result_json) == 1
+
+        return data.LocalIdentity(**result_json[0])
+
+    async def _get_registration_challenge(
+        self,
+        email: str,
+        user_handle: bytes,
+    ) -> WebAuthnRegistrationChallenge:
+        result = await execute.parse_execute_json(
+            self.db,
+            """
+select ext::auth::WebAuthnRegistrationChallenge {
+    id,
+    challenge,
+    user_handle,
+    email,
+}
+filter .email = <str>$email
+and .user_handle = enc::base64_decode(<str>$user_handle);""",
+            variables={
+                "email": email,
+                "user_handle": base64.b64encode(user_handle).decode(),
+            },
+            cached_globally=True,
+        )
+        result_json = json.loads(result.decode())
+        assert len(result_json) == 1
+        challenge_dict = result_json[0]
+
+        return WebAuthnRegistrationChallenge(
+            id=challenge_dict["id"],
+            challenge=base64.b64decode(challenge_dict["challenge"]),
+            user_handle=base64.b64decode(challenge_dict["user_handle"]),
+            email=challenge_dict["email"],
+        )
+
+    async def _delete_registration_challenges(
```

Comment:
Should we also do cleanup of old challenges, like we do for pkce?

Reply:
Since we delete all challenges after a successful registration which means the only "old challenges" we would have are ones where someone began the WebAuthn process and then decided not to use WebAuthn. I guess that's a valid case for cleaning up old challenges but that's far less likely to happen than abandoned/timed-out PKCE flows.

I can look at adding that as a separate PR maybe?